File: src/backup-tool-svc.js
/**
 * BackupToolService class provides methods to take snapshots of directories,
 * list snapshots, restore snapshots, prune snapshots, and check the integrity
 * of the database.
 */
const fs = require('fs');
const path = require('path');
const Table = require('cli-table3');
const crypto = require('crypto');

const { TECHNIQUE } = require('./technique')
const BackupToolDAO = require('./backup-tool-dao');



class BackupToolService {
  /**
   * Initalizes the BackupToolService with a new BackupToolDAO instance. 
   * and configuration values for the backup technique and chunk size.
   */
  constructor() {
    this.dao = new BackupToolDAO();
    this.technique = process.env.TECHNIQUE;
    this.chunkSize = process.env.CHUNK_SIZE;
  }

  /**
   * Takes a snapshot of the target directory and stores the snapshot in the database.
   * @param {*} targetDirectory The directory to take a snapshot of. 
   * @returns {Promise<number>} The ID of the snapshot that was taken.
   */
  async takeSnapshot(targetDirectory) {
    const timestamp = new Date().toISOString();
    const directorySize = await this.#getDirectorySize(targetDirectory);
    const snapshotId = await this.dao.insertSnapshot(timestamp, directorySize, this.technique);

    const files = fs.readdirSync(targetDirectory);
    for (const file of files) {
      const filePath = path.join(targetDirectory, file);
      const fileSize = fs.statSync(filePath).size;
      if (this.technique === TECHNIQUE.CHUNKING) {
        await this.#snapshotByChunk(snapshotId, filePath, fileSize);
      } else {
        await this.#snapshotByFile(snapshotId, filePath, fileSize);
      }
    }
    return snapshotId;
  }


  /**
   *  Lists all snapshots stored in the database and prints them to the console in a table format
   * @returns {Promise<Array>} A list of snapshots stored in the database.
   */
  async listSnapshots() {
    const snapshots = await this.dao.listSnapshots();
    let table = new Table({
      head: ['SNAPSHOT', 'TIMESTAMP', 'DIRECTORY SIZE', 'SNAPSHOT SIZE', 'TECHNIQUE']
    });
    for (const snapshot of snapshots) {
      const snapshotSize = await this.dao.getSnapshotSize(snapshot.id);
      table.push(
        [snapshot.id, new Date(snapshot.timestamp).toLocaleString(), snapshot.directory_size, snapshotSize, snapshot.technique]
      );
    }
    console.log(table.toString());
    const totalDbSize = await this.dao.getTotalDatabaseSize();
    console.log(`Total database size: ${totalDbSize}`);
    return snapshots;
  }

  /**
   * Restores a snapshot by recreating the files in the output directory.
   * @param {*} snapshotId 
   * @param {*} outputDirectory 
   */
  async restoreSnapshot(snapshotId, outputDirectory) {
    const snapshot = await this.dao.getSnapshot(snapshotId);
    if (snapshot) {
      const files = await this.dao.getFilesBySnapshotId(snapshotId);
      for (const file of files) {
        const filePath = path.join(outputDirectory, file.path);
        if (snapshot.technique === TECHNIQUE.CHUNKING) {
          await this.#restoreFromChunk(file, filePath);
        } else {
          await this.#restoreFromContent(file, filePath);
        }
      }
    }
  }

  /**
   * Deletes a snapshot, files, and file_chunk(if applicable) from the database.
   * @param {*} snapshotId 
   */
  async pruneSnapshots(snapshotId) {
    const snapshot = await this.dao.getSnapshot(snapshotId);
    if (snapshot) {
      await this.dao.deleteSnapshot(snapshotId);
    }
  }

  /**
   * Checks the database for corrupted file content by comparing the hash of the content
   */
  async checkDatabase() {
    const fileContents = await this.dao.getAllFileContents();
    let corruptedFiles = [];

    for (const file of fileContents) {
      const computedHash = await this.#hashContent(file.content);
      if (computedHash !== file.hash) {
        corruptedFiles.push(file.hash);
      }
    }
    if (corruptedFiles.length > 0) {
      console.log('Corrupted files found:');
      corruptedFiles.forEach(hash => console.log(`Hash: ${hash}`));
    } else {
      console.log('No corrupted files found.');
    }
  }

  /**
   * ********** PRIVATE METHODS **********
   */

  /**
   * Takes a snapshot of a file by chunking the file into smaller chunks
   * @param {*} snapshotId 
   * @param {*} filePath 
   * @param {*} fileSize 
   */
  async #snapshotByChunk(snapshotId, filePath, fileSize) {
    const fileId = await this.dao.insertFile(snapshotId, filePath, '', fileSize); // Empty hash for now
    const fileChunks = await this.#chunkFile(filePath);

    for (let i = 0; i < fileChunks.length; i++) {
      const chunk = fileChunks[i];
      const chunkHash = await this.#hashContent(chunk);

      // Check if the chunk hash already exists
      const existingChunk = await this.dao.getChunkByHash(chunkHash);
      if (!existingChunk) {
        // If the chunk hash does not exist, store the chunk
        await this.dao.insertChunk(chunkHash, chunk);
      }
      // Store the file chunk reference
      await this.dao.insertFileChunk(fileId, chunkHash, i);
    }
  }
  /**
   * Takes a snapshot of a file by storing the file content directly
   * @param {*} snapshotId 
   * @param {*} filePath 
   * @param {*} fileSize 
   */
  async #snapshotByFile(snapshotId, filePath, fileSize) {
    const fileHash = await this.#hashFile(filePath);
    const content = await this.dao.getContentByHash(fileHash);
    if (!content) {
      // If the content hash does not exist, store the content
      const fileContent = fs.readFileSync(filePath);
      await this.dao.insertContent(fileHash, fileContent);
    }
    // Store the file path, hash, and size in the files table
    await this.dao.insertFile(snapshotId, filePath, fileHash, fileSize);

  }
  /**
   * Restores a file from its content
   * @param {*} file 
   * @param {*} filePath 
   */
  async #restoreFromContent(file, filePath) {
    const fileContent = await this.dao.getContentByHash(file.hash);
    fs.mkdirSync(path.dirname(filePath), { recursive: true });
    fs.writeFileSync(filePath, fileContent);
  }

  /**
   * Restores a file from its chunks
   * @param {*} file 
   * @param {*} filePath 
   */
  async #restoreFromChunk(file, filePath) {
    const fileChunks = await this.dao.getFileChunks(file.id);
    const fileBuffer = Buffer.concat(await Promise.all(fileChunks.map(async (chunk) => {
      return await this.dao.getChunkByHash(chunk.chunk_hash);
    })));
    fs.mkdirSync(path.dirname(filePath), { recursive: true });
    fs.writeFileSync(filePath, fileBuffer);
  }

  async #hashFile(filePath) {
    const fileBuffer = fs.readFileSync(filePath);
    return this.#hashContent(fileBuffer);
  }

  async #hashContent(content) {
    const hash = crypto.createHash('sha256');
    hash.update(content);
    return hash.digest('hex');
  }

  async #chunkFile(filePath, chunkSize = 1024 * 1024) { // 1MB chunks
    const fileBuffer = fs.readFileSync(filePath);
    const chunks = [];
    for (let i = 0; i < fileBuffer.length; i += chunkSize) {
      chunks.push(fileBuffer.subarray(i, i + chunkSize));
    }
    return chunks;
  }

  async #getDirectorySize(directory) {
    const files = fs.readdirSync(directory);
    let totalSize = 0;
    for (const file of files) {
      const filePath = path.join(directory, file);
      const stats = fs.statSync(filePath);
      totalSize += stats.size;
    }
    return totalSize;
  }
}

module.exports = BackupToolService;

File: python_src/backup-tool-svc.py
import os
import hashlib
import datetime
import argparse
from pathlib import Path

from cli_table import Table
from backup_tool_dao import BackupToolDAO
from technique import TECHNIQUE  # Import the TECHNIQUE enum

class BackupToolService:
    def __init__(self):
        self.dao = BackupToolDAO()
        self.technique = TECHNIQUE[os.getenv('TECHNIQUE', TECHNIQUE.CHUNKING.value)]
        self.chunk_size = int(os.getenv('CHUNK_SIZE', 1024 * 1024))  # Default to 1MB

    def take_snapshot(self, target_directory):
        timestamp = datetime.datetime.now().isoformat()
        directory_size = self.get_directory_size(target_directory)
        snapshot_id = self.dao.insert_snapshot(timestamp, directory_size, self.technique.value)

        for root, _, files in os.walk(target_directory):
            for file in files:
                file_path = os.path.join(root, file)
                file_size = os.path.getsize(file_path)
                if self.technique == TECHNIQUE.CHUNKING:
                    self.snapshot_by_chunk(snapshot_id, file_path, file_size)
                else:
                    self.snapshot_by_file(snapshot_id, file_path, file_size)
        return snapshot_id

    def list_snapshots(self):
        snapshots = self.dao.list_snapshots()
        table = Table(['SNAPSHOT', 'TIMESTAMP', 'DIRECTORY SIZE', 'SNAPSHOT SIZE', 'TECHNIQUE'])
        for snapshot in snapshots:
            snapshot_size = self.dao.get_snapshot_size(snapshot['id'])
            table.add_row([snapshot['id'], snapshot['timestamp'], snapshot['directory_size'], snapshot_size, snapshot['technique']])
        print(table)
        total_db_size = self.dao.get_total_database_size()
        print(f'Total database size: {total_db_size}')
        return snapshots

    def restore_snapshot(self, snapshot_id, output_directory):
        snapshot = self.dao.get_snapshot(snapshot_id)
        if snapshot:
            files = self.dao.get_files_by_snapshot_id(snapshot_id)
            for file in files:
                file_path = os.path.join(output_directory, file['path'])
                if snapshot['technique'] == TECHNIQUE.CHUNKING.value:
                    self.restore_from_chunk(file, file_path)
                else:
                    self.restore_from_content(file, file_path)

    def prune_snapshots(self, snapshot_id):
        snapshot = self.dao.get_snapshot(snapshot_id)
        if snapshot:
            self.dao.delete_snapshot(snapshot_id)

    def get_directory_size(self, directory):
        total_size = 0
        for dirpath, dirnames, filenames in os.walk(directory):
            for f in filenames:
                fp = os.path.join(dirpath, f)
                total_size += os.path.getsize(fp)
        return total_size

    def snapshot_by_chunk(self, snapshot_id, file_path, file_size):
        with open(file_path, 'rb') as f:
            chunk_number = 0
            while chunk := f.read(self.chunk_size):
                chunk_hash = hashlib.sha256(chunk).hexdigest()
                self.dao.insert_chunk(snapshot_id, file_path, chunk_number, chunk_hash, chunk)
                chunk_number += 1

    def snapshot_by_file(self, snapshot_id, file_path, file_size):
        with open(file_path, 'rb') as f:
            content = f.read()
            file_hash = hashlib.sha256(content).hexdigest()
            self.dao.insert_file(snapshot_id, file_path, file_hash, content)

    def restore_from_chunk(self, file, file_path):
        os.makedirs(os.path.dirname(file_path), exist_ok=True)
        with open(file_path, 'wb') as f:
            chunks = self.dao.get_chunks_by_file(file['id'])
            for chunk in chunks:
                f.write(chunk['data'])

    def restore_from_content(self, file, file_path):
        os.makedirs(os.path.dirname(file_path), exist_ok=True)
        with open(file_path, 'wb') as f:
            f.write(file['content'])

def main():
    backup_tool_service = BackupToolService()

    parser = argparse.ArgumentParser(description='Backup Tool')
    subparsers = parser.add_subparsers(dest='command')

    snapshot_parser = subparsers.add_parser('snapshot', help='Take a snapshot of a directory')
    snapshot_parser.add_argument('targetDirectory', type=str, help='Target directory to snapshot')

    list_parser = subparsers.add_parser('list', help='List all snapshots')

    restore_parser = subparsers.add_parser('restore', help='Restore a snapshot')
    restore_parser.add_argument('snapshot', type=int, help='Snapshot ID to restore')
    restore_parser.add_argument('outputDirectory', type=str, help='Output directory for restored files')

    prune_parser = subparsers.add_parser('prune', help='Prune old snapshots')
    prune_parser.add_argument('snapshot', type=int, help='Snapshot ID to prune')

    args = parser.parse_args()

    if args.command == 'snapshot':
        backup_tool_service.take_snapshot(args.targetDirectory)
    elif args.command == 'list':
        backup_tool_service.list_snapshots()
    elif args.command == 'restore':
        backup_tool_service.restore_snapshot(args.snapshot, args.outputDirectory)
    elif args.command == 'prune':
        backup_tool_service.prune_snapshots(args.snapshot)
    else:
        parser.print_help()

if __name__ == '__main__':
    main()